tools/bin/commands/link.ts

```diff
@@ -1,175 +1,133 @@
-        // Insert the internal dependency before the main package
-        hierarchy.splice(idx, 0, internalDep);
-        idx++;
-      } else if (depIdx > idx) {
-        // Remove the internal dependency from the list and place it before the main package
-        // (doesn't affect the order of same-level libraries, their order is arbitrary)
-        hierarchy.splice(depIdx, 1);
-        idx = hierarchy.indexOf(dep);
-        hierarchy.splice(idx, 0, internalDep);
-        idx++;
-      }
-    }
+function parsePackageDependencies(dependencyName: string, pathToLink: string): string[] {
+  let result: string[] = [];
+  if (!packagesToLink.has(dependencyName)) {
+    packagesToLink.add(dependencyName);
+    localPackageDependencies.push(new PackageData(dependencyName, pathToLink));
   }
-  return hierarchy;
+  result.push(dependencyName);
+  return result;
 }
 
-/** Executes a script for a package and its dependencies with messages on the current progress. */
-function runScript(packageDir: string, script: string, dependencies: utils.Indexable, options:
-  {dirMessage?: string, successMessage?: string, callback?: (dir: string) => void} = {}) {
-  exec(script, {cwd: packageDir}, (err, stdout, stderr) => {
-    if (options.dirMessage)
-      console.log(`${options.dirMessage}${path.basename(packageDir)}`);
-    if (options.callback)
-      options.callback(packageDir);
-    if (err) throw err;
-    else {
-      console.log(stderr, stdout);
-      for (const dep in dependencies) {
-        const depPath = getPackagePath(dep);
-        exec(script, {cwd: depPath}, (err, stdout, stderr) => {
-          if (options.dirMessage)
-            console.log(`${options.dirMessage}${path.basename(depPath)}`);
-          if (options.callback)
-            options.callback(depPath);
-          if (err) throw err;
-          else console.log(stderr, stdout);
-        });
-      }
-      // if (options.successMessage)
-      //   setTimeout(() => color.success(options.successMessage!), 5000);
+
+async function linkPackages() {
+  let anyChanges = true;
+
+  while (anyChanges && packagesToLink.size > 0) {
+    anyChanges = false;
+    let mapElements = localPackageDependencies.filter(x => x.dependencies.every(i => !packagesToLink.has(i)) && packagesToLink.has(x.name));
+
+    for (let element of mapElements) {
+      await runScript(`npm install`, element.packagePath);
+      for (let dependency of element.dependencies)
+        await runScript(`npm link ${dependency}`, element.packagePath);
+      await runScript(`npm link`, element.packagePath);
+      packagesToLink.delete(element.name);
+      anyChanges = true;
     }
-  });
+
+    if (anyChanges === false)
+      throw (new Error(`There is loop with next packages: ${JSON.stringify(Array.from(packagesToLink)).toString()}`));
+  }
+
+  let names = localPackageDependencies.map(x => x.name);
+  await runScript(`npm install`, curDir);
+  for (let name of names) {
+    await runScript(`npm link ${name}`, curDir);
```

Comment:
replace this with 
`await runScript('npm link ${names.join(' ')}', curDir);`

what you have can cause weird behavior by linking only last package, sometimes npm does weird things